pypit/core/arextract.py

```diff
@@ -30,6 +30,934 @@
+    """ Extract the total flux within a boxcar window at many positions.
+    This routine will accept an asymmetric/variable window
+    Traces are expected to run vertically to be consistent with other
+    extract_  routines. Based on idlspec2d/spec2d/extract_asymbox2.pro
+
+    Parameters
+    ----------
+    image :   numpy float 2-d array [nspec, nspat]
+    left  :   Lower boundary of boxcar window (given as floating pt pixels) [nspec, nTrace]
+    right     - Upper boundary of boxcar window (given as floating pt pixels) [nspec, nTrace]
+
+    Optional Parameters
+    -------------------
+    ycen :    Y positions corresponding to "left" (expected as integers) [nspec,nTrace]
+    weight_image:  Weights to be applied to image before boxcar [nspec, nspat]
+
+    Returns
+    -------
+    fextract:   Extracted flux at positions specified by (left<-->right, ycen) [nspec, nTrace]
+
+    Revision History
+    ----------------
+    24-Mar-1999  Written by David Schlegel, Princeton.
+    17-Feb-2003  Written with slow IDL routine, S. Burles, MIT
+    27-Jan-2004  Adopted to do asymmetric/varying boxcar
+    22-Apr-2018  Ported to python by Joe Hennawi
+    """
+
+    left = left_in.T
+    right = right_in.T
+
+    dim = left.shape
+    ndim = left.ndim
+    if (ndim == 1):
+        nTrace = 1
+        npix = dim[0]
+    else:
+        nTrace = dim[0]
+        npix = dim[1]
+
+    if ycen == None:
+        if ndim == 1:
+            ycen = np.arange(npix, dtype='int')
+        elif ndim == 2:
+            ycen = np.outer(np.ones(nTrace, dtype='int'), np.arange(npix, dtype='int'), )
+        else:
+            raise ValueError('left is not 1 or 2 dimensional')
+
+    ycen_out = ycen.astype(int)
+    if np.size(left) != np.size(ycen):
+        raise ValueError('Number of elements in left and ycen must be equal')
+
+    idims = image.shape
+    nspat = idims[1]
+    nspec = idims[0]
+
+    maxwindow = np.max(right - left)
+    tempx = np.int(maxwindow + 3.0)
+
+    bigleft = np.outer(left[:], np.ones(tempx))
+    bigright = np.outer(right[:], np.ones(tempx))
+    spot = np.outer(np.ones(npix * nTrace), np.arange(tempx)) + bigleft - 1
+    bigy = np.outer(ycen_out[:], np.ones(tempx, dtype='int'))
+
+    fullspot = np.array(np.fmin(np.fmax(np.round(spot + 1) - 1, 0), nspat - 1), int)
+    fracleft = np.fmax(np.fmin(fullspot - bigleft, 0.5), -0.5)
+    fracright = np.fmax(np.fmin(bigright - fullspot, 0.5), -0.5)
+    del bigleft
+    del bigright
+    bool_mask1 = (spot >= -0.5) & (spot < (nspat - 0.5))
+    bool_mask2 = (bigy >= 0) & (bigy <= (nspec - 1))
+    weight = (np.fmin(np.fmax(fracleft + fracright, 0), 1)) * bool_mask1 * bool_mask2
+    del spot
+    del fracleft
+    del fracright
+    bigy = np.fmin(np.fmax(bigy, 0), nspec - 1)
+
+    if weight_image != None:
+        temp = np.array([weight_image[x1, y1] * image[x1, y1] for (x1, y1) in zip(bigy.flatten(), fullspot.flatten())])
```

Comment:
This would be a lot easier to read if written:

```
for (x1,y1) in zip(...):
    np.array([weight_image[x1,y1]*image[x1,y1])
```

Same goes throughout the code.

Reply:
Oh, I see now.  You are doing a list iterator.  Nevermind.